Feature: LoRaWAN TTN controller  RN2384 [Experimental]

TD-er opened the pull request:
A very experimental new controller to add LoRaWAN support to ESPeasy.
The data is not yet tightly packed. It is just a dump of data from the sensor sending it to the controller.
The only optimization is to send only the number of values a plugin outputs.

The values are currently  converted to uint32_t by multiplying with 10'000.
On the receiving end the values are decoded back to (signed) floats with 4 decimals resolution.

Payload: `027B000200E20400B0E9F7FF`
Decoded:
```json
{
  "IDX": 123,
  "taskindex": 2,
  "val_1": 32,
  "val_2": -53,
  "valuecount": 2
}
```

Datarate is currently fixed at SF7/BW125
Even the GPIO pins and baud rate is currently fixed for my test setup.
Right now only OTAA is supported, but to support ABP I have to change some of the controller parameter definitions. (which also may lead to longer PW/user for other controllers)

The used LoRa chip is a Microchip RN2384.

De decode.js file is borrowed from https://github.com/cyberman54/ESP32-Paxcounter, but I've been working on that one too :) https://github.com/cyberman54/ESP32-Paxcounter/pull/405

TD-er committed:
[TTN] Add RN2483-Arduino-Library

TD-er committed:
[TTN] Skeleton for C018 TTN RN2xx3 controller

Initial support for LoRaWan TTN controller for ESPeasy

TD-er committed:
Improve RN2xx3 LoRo TTN library

TD-er committed:
[LoRa TTN] Add some statistics to the controller info page

TD-er committed:
[RN2384 lib] Only use const String& in function parameters

Making these functions a lot more memory efficient

TD-er committed:
[LoRa plugin] Use separate object for C018 data

To avoid object initialization before the controller is set active.

TD-er committed:
[LoRa TTN] Create special interface for LoRa TTN controller

Lots of elements of a standard controller are not needed for LoRa, or called differently.

TD-er committed:
[Controller page] Restructure Controller rendering page

Move loading and saving of controller settings to separate functions per item.
Also string generation is now done in a function to keep internal strings in sync.

TD-er committed:
Missing break; statement in controllers function switch statement

Some controllers were missing a break; in switch statements.
This would then also execute the next case in the switch list.

TD-er committed:
[Cleanup] Split Webserver.ino is separate files per page

TD-er committed:
[Cleanup] Uncrustify of all WebServer_*.ino files

TD-er committed:
[Cleanup] Split handle_controllers() function into separate functions

The handle_controllers function is rather long, so it is now split into one main wrapper and 4 smaller parts.

TD-er committed:
[Cleanup] Split handle_devices() function into separate functions

TD-er committed:
[Cleanup] Stop using global declared dummyString

This string may end up keeping quite some memory for way longer than actually needed.

TD-er committed:
[LoRa TTN] Show Dev Addr in controller overview page as host

Normally a controller uses a hostname or IP to connect to.
But a LoRa node does not connect to any specific host, but does broadcast messages using its own unique dev addr.
This will be shown as host description in the Controller overview page.

TD-er committed:
[LoRa TTN] Send data to the TTN network + decoder

TD-er commented:
A lot of other files also had to be changed, in order to get some overview over the WebServer messy code.
So all pages are now split over several files and a few small bugs were fixed while cleaning up this code :)

The decoded data on the TTN console:
![image](https://user-images.githubusercontent.com/3751318/62815710-badc7800-bb1c-11e9-8cd4-11655502476c.png)


The new TTN controller page:
![image](https://user-images.githubusercontent.com/3751318/62815727-10188980-bb1d-11e9-82f3-0c9bd401f57d.png)

LoRaWAN dev addr in controller overview page:
![image](https://user-images.githubusercontent.com/3751318/62815741-435b1880-bb1d-11e9-99be-bf04468a2976.png)

ESPeasy p2p shared plugin is now also easier to see its source:
![image](https://user-images.githubusercontent.com/3751318/62815765-7b625b80-bb1d-11e9-837d-3992bd042df1.png)

TD-er committed:
[LoRa TTN] Start working on the interpreter of decoded data

Currently BME280 and SystemInfo are being detected.

TD-er committed:
[Encoder] ICACHE_RAM_ATTR added for encoder plugin

Fixes #2492

TD-er committed:
[LoRa TTN] Add basic decoder for all plugins

Next step will be to let the plugin generate the binary data.

micropet commented:
Wau!

But where do you see a purpose for it?
I have Wi-Fi indoor and can send my measurements to the internet.

Interesting would be a Wemos D1 with battery and a maturity of several months.

Do you have a link to a compiled version?

TD-er commented:
Not yet, but I will make one later for you. (GPIO's are not selectable yet)
Do you have the used RN2384 module?

My purpose is to have sensors placed in "the field" or for mobile use.
Think about festivals, where WiFi coverage is limited, or agricultural use where you literally have sensors "in the field".

And for [the event we (Jimmy and I) had last May](http://www.delsboelectric.se/), where we had to take measurements along a track of over 3 km.

micropet commented:
Yes, I have a RN2384 module and an IC880 gateway here. The things were not longer in operation. I will connect the module once.

micropet commented:
can see right now: my module is called RN2483A
Do you mean that?

So, my gateway is connected and the module is connected to a Wemos D1.
If you can give me the code for Platformio or a compiled file, we can start.

TD-er committed:
[Docs] Building ESPeasy using PlatformIO in VS-Code

As discussed on the [Forum](https://www.letscontrolit.com/forum/viewtopic.php?f=6&t=6922&p=38512#p38512)

TD-er commented:
I just made a [test build](https://www.dropbox.com/s/xssi06whnqcsiur/custom_ESP8266_4M_LoRa_TTN_testbuild.rar?dl=0) for it.

Please note that the GPIO pins are not configurable right now, so you have to use:
- int rxpin    = 12;
- int txpin    = 14;

Reset pin is also not yet implemented, but I have not yet had to use it.
The decoder and converter JavaScript files:
- [Decoder](
https://github.com/letscontrolit/ESPEasy/blob/1d2c3ef87b4ed4ee244f03b1d80bee7ca9509096/misc/TTN/packed_decoder.js)
- [Converter](https://github.com/letscontrolit/ESPEasy/blob/1d2c3ef87b4ed4ee244f03b1d80bee7ca9509096/misc/TTN/packed_converter.js)

Those last 2 ones I am now changing, since it is not really practical to have it like this. Some plugins do allow to select a value for output and then we need to guess at the TTN side what value it is.
So I will add a "raw" format to all plugins to make sure they always send values in the same order.
This also allows me to compact the data stream, since the plugin knows best what range and resolution a value must have.

Let me know if you need more info or run into bugs.

micropet commented:
It takes a while.
When I enter the controller and submit, the wemos hangs.

TD-er commented:
Then it doesn't see the sensor.
Please check the used GPIO pins.
You probably swapped them :) (happened here too)

micropet commented:
I flashed again. The data on the controller side is correct.
Then I pressed submit and already the Wemos is hanging.

I check the RX and TX lines.

TX from the module goes to D6
RX from the module goes to D5

micropet commented:
Could narrow the problem a bit:

I took the values from a Thermo Switch.
But that is only in the testing version in it.

TD-er commented:
You should at least see some info like this, if the module is detected:
![image](https://user-images.githubusercontent.com/3751318/62839278-5b02e000-bc88-11e9-86a9-c6eda903a3b1.png)

If you don't see that, you cannot do anything with it.
Well at least the Version Number and the DevEUI should be visible and perhaps also the voltage.


Edit: Just to be sure, I flashed exactly the same build as you're having and it is working here.
Just to be sure, you may want to swap the RX/TX lines. I know I did mess them up the first time here also, so maybe there is something wrong in my code :)

micropet commented:
![TTN1](https://user-images.githubusercontent.com/2838584/62839383-9ce05600-bc89-11e9-87a0-551ac98d362d.JPG)

micropet commented:
This does not look good.
Unfortunately, I have not connected any sensors that work with the "Normal" build

For each action - click Device Controllers or other pages, the wemos hangs for 5 to 7 minutes.

Hast Du auch eine "Testing" Version da?

![TTN2](https://user-images.githubusercontent.com/2838584/62839441-0f513600-bc8a-11e9-8c81-d7e9a9efa975.JPG)

TD-er commented:
I can also make a 'normal' build for you with this controller included.

["Normal" Test Build](https://www.dropbox.com/s/h51vdqgdqjpefxg/normal_ESP8266_4M_LoRa_testbuild.rar?dl=0)

What did you have to change to get the module recognized?
And do you see the module perform the OTAA in the application page on the Things Network console?
![image](https://user-images.githubusercontent.com/3751318/62839726-71ac3580-bc8e-11e9-859d-1bddc80bbcd7.png)

micropet commented:
I need a "Testing" Build

micropet commented:
I have the controller 2 on the dummy now.
Looks good, right?
But he does not send any data.

![TTN3](https://user-images.githubusercontent.com/2838584/62839731-b20bb380-bc8e-11e9-8e9d-1d4d3535322e.JPG)

TD-er commented:
You should see the Dev Addr mentioned in the controllers page like this:
![image](https://user-images.githubusercontent.com/3751318/62839739-e1babb80-bc8e-11e9-8839-62d49750775f.png)

If you don't see that, it does not successfully perform an OTAA.
Either the gateway cannot be reached, or the answer cannot be heard on the module.
I have not yet implemented ABP, so if for whatever reason your module cannot receive (I have one such module here), you cannot perform OTAA.

At least the request should be seen on the Gateway console page as incoming data.

TD-er commented:
A ["Test" build for LoRa TTN](https://www.dropbox.com/s/rcagf0j2iruxlwe/test_core_252_ESP8266_4M_LoRa_testBuild.rar?dl=0)

micropet commented:
The mistake is mine. My gateway is suddenly disconnected.

TD-er commented:
Please also have a look at the web pages to see if something is not working like it should.
I did move all pages into separate files and I redid the implementation of the Controller page.
This way it is a lot easier to navigate through the close to 10k lines of code regarding the webserver
That was a long overdue split of code.

Also the code for the controller page is now actually readable and I also split the plugin page into 5 functions, one for a different stage of the page.

micropet commented:
I had plugged the IC880 Gateway into a socket that only has voltage as long as the motion sensor in the basement turns on the light. :)

micropet commented:
I think I did everything right now.
The TTN controller is on a thermo sensor.
But in the TTN no data arrive.

TD-er commented:
In the controller page you should see how many packets it already sent.
And just to be sure, set the minimum interval to something like 10 sec, to make sure you're not hitting the 1% sending time.

Checklist:
- You see a Dev Addr
- Uplink Frame Counter is not 0
- No reported errors in the Controller page

N.B. the controller page data is only rendered on loading the page.
Do not refresh after submitting, but navigate to it again.

As a test you can remove the decoder/converter scripts. You will then see the raw payload in the application console page on ThingsNetwork

micropet commented:
No Dev Addr
Uplink Counter is 0
No Errors

TTN says "never seen"

TD-er commented:
Then the OTAA is not working well.
Please check the frame counter check to be disabled in TTN console page and that you have the right OTAA values (No bytes are missing in the field due to truncating)

Oh and just to be sure, I am using the EU_868 parameters.

micropet commented:
frame counter check is disabled 

I thing i use the EU_868 parameters. 
2 years ago I had some battery powered devices running. That was always OK.

TD-er commented:
Do you see the gateway on your account?
Can you see any data there? (not only this device)

micropet commented:
I added the Application EUI and App Key again.
But still does not work.

The gateway says connected

![TTN4](https://user-images.githubusercontent.com/2838584/62840193-b5a33880-bc96-11e9-80c7-c2193bff346b.JPG)

TD-er commented:
Did you power the ESP node off and on again?
Like I said, I do not yet have the reset of the RN2384 implemented.

micropet commented:
Yes, because it hangs over and over again, I turned it off and on very often.

micropet commented:
I'll try another RN2483 module tomorrow.
I go to sleep now and continue tomorrow.
Many thanks for your effort.

TD-er committed:
[LoRa TTN] Add decoder types

micropet commented:
good Morning.

My gateway is fine.
I have an old node in operation and it sends the data to TTN Network.

The Wemos still does not work.
I ordered a new module from GB on ebay.

That means, at the moment I'm stuck.

![TTN Node62 Data](https://user-images.githubusercontent.com/2838584/62856152-6c81d180-bcf4-11e9-8d88-044a5e8ef288.JPG)

TD-er commented:
I made a new ["test" test build](https://www.dropbox.com/s/m4i1mn31zf1u4gu/test_core_252_ESP8266_4M_LoRa_testBuild_errorOTAA.rar?dl=0)
This one also keeps track of the returned text during OTAA if it is not "accepted".
Maybe it will help on finding the issue here.
Please check or try to find out if your gateway and/or connection profile should allow
```
  "data_rate": "SF7BW125",
  "coding_rate": "4/5",
```

micropet commented:
I flashed it.
TTN says "never seen"

How can I find out if my gateway can do that?

And where can I see the message "OTAA if it is not accepted"?

TD-er commented:
I just made [another build](https://www.dropbox.com/s/6jwfw57a5tw8tjm/test_core_252_ESP8266_4M_LoRa_testBuild_errorOTAA_setDR.rar?dl=0)
This one does explicit set the datarate.

The error should be shown at the bottom on the controller settings page.
Where you also see the Uplink Frame Counter etc.

micropet commented:
OK, flashed the last image.
Here the controller side:

![TTN5](https://user-images.githubusercontent.com/2838584/62860709-5c242380-bd01-11e9-8edc-b71b4ec76f6e.JPG)

TD-er commented:
Hmm by reading it, the error is cleared. Maybe that's not the best design choice here.
Let me change that.
One moment

micropet commented:
OK

TD-er commented:
Please check [test_core_252_ESP8266_4M_LoRa_testBuild_NoClearError.rar](https://www.dropbox.com/s/139asyewhkqjvms/test_core_252_ESP8266_4M_LoRa_testBuild_NoClearError.rar?dl=0)

micropet commented:
Something is not normal, the Wemos always needs 3 to 4 minutes to start.

![ttn6](https://user-images.githubusercontent.com/2838584/62861262-2d0eb180-bd03-11e9-8946-20df4c1a9726.JPG)

TD-er commented:
On my test node, I do see the LoRa join on the TTN console before I even can load the web page of it.
So that should be matter of seconds here. (14 max according to the timing stats)

[Build keeping all error state strings](https://www.dropbox.com/s/nr7z0m9bt43eonl/test_core_252_ESP8266_4M_LoRa_testBuild_KeepAllErrors.rar?dl=0)
Well not all, but I do keep the returned string if something is wrong, so now you should see some error.... I hope.
I will also add some logging, since apparently we need it here :)

micropet commented:
I also think that logging is important :)
Yes, weird after the flashes or reset the Wemos needs 4 minutes to see the web pages.

![TTN7](https://user-images.githubusercontent.com/2838584/62861764-c8eced00-bd04-11e9-87c8-9383e1732176.JPG)

micropet commented:
![TTN8](https://user-images.githubusercontent.com/2838584/62861903-21bc8580-bd05-11e9-942e-1d1e8dad1f14.JPG)

TD-er commented:
The library I used is blocked waiting and it does retry a few times incl. long wait times.
That's one thing that has to be changed, but like I said it is still in the early stages

micropet commented:
Yes, good.

Maybe my module is broken. It was brand new. I connected it yesterday for the first time.
Next days, the new module from GB would have to arrive here.

I had previously worked with the RFM95, the RN2483 was always too expensive for me and is just awkward to get.

micropet commented:
I will have lunch now.
Maybe you have an idea ...

TD-er commented:
I will add ABP join, so you can see if that's working.
If so, then I suspect it to be unable to receive.

And the RN2384 is indeed a bit more expensive, but not that much anymore.
Farnell has them for about 11 euro ex. VAT.
They are a bit hard to solder, but I like the fact they have the entire stack implemented. So that takes away the time critical part.
For ESP32 the RFM95 or SX1276 can be used, since that CPU does have more clock cycles to spare (you could run it on a dedicated core).
Adding support for those will be the next step.

TD-er commented:
A new build. Error logging when trying to join and no TX when not joined.
[test_core_252_ESP8266_4M_LoRa_OnlyTX_joined](https://www.dropbox.com/s/pp5ox4onwktocj8/test_core_252_ESP8266_4M_LoRa_OnlyTX_joined.rar?dl=0)

This should speedup your boot and also the "not_joined" error is a return value when trying to send data.
By keeping track of the joined state, the TX command should not be given to start sending data.
When seeing this error, the library does call again for an init and does retry to send data....
The timeouts for this are enormous.

micropet commented:
Well, just flashed.

There is nothing to see in the log.
The Controller page does not show an error.
But is not connected.

Do I have to switch to ABP at TTN now?

For ABP I have to register the Network Session Key and an App Session Key.

TD-er commented:
No, I have not finished adding ABP support.

TD-er commented:
The node you have that is able to connect.
Do you know if you set it to SF8, or is it forced to that by the OTAA?

TD-er commented:
Hmm, I noticed I never call to set the frequency plan.
Will now make a version that does set the frequency plan.

micropet commented:
// Set data rate and transmit power (note: txpow seems to be ignored by the library)
    LMIC_setDrTxpow(DR_SF8,14);  // Airtime BW125 SF7 = 82ms SF8 = 160ms  SF9 = 160ms SF11 = 1050ms SF12 = 1810ms 
    
    // uses SF9 for its RX2 window
    LMIC.dn2Dr = DR_SF9; // für den Empfang
/*
    // Use a frequency in the g3 which allows 10% duty cycling.
    LMIC.freq = 869525000;
    // Maximum TX power
    LMIC.txpow = 27;
    // Use a medium spread factor. This can be increased up to SF12 for
    // better range, but then the interval should be (significantly)
    // lowered to comply with duty cycle limits as well.
    LMIC.datarate = DR_SF9;
    // This sets CR 4/5, BW125 (except for DR_SF7B, which uses BW250)
    LMIC.rps = updr2rps(LMIC.datarate);


n the Arduino code I think SF8 is set.
It's been 2 years since I've dealt with it. I already forgot everything.

TD-er commented:
[Build with set frequencyplan](https://www.dropbox.com/s/gqj8q2h1hm9ii34/test_core_252_ESP8266_4M_LoRa_testBuild_setFreqPlan.rar?dl=0)

micropet commented:
You are fast!

Now the Wemos boots quite fast.
But no connect.

TD-er commented:
and no error?

micropet commented:
No Error. No Connect.

TD-er commented:
That's very strange. 
Like the sensor freezes and does not return anything.

I will add some settings so you can test SF, OTAA/ABP, etc

You don't see anything coming in on the gateway?
Maybe double check to see if the antenna is connected to the right pin? (there are 2 RF pins)

micropet commented:
My working node says:

   "data_rate": "SF8BW125",
   "coding_rate": "4/5",

I've been testing everything since last night.

The antenna is also on the 868 MHz pin.

A tail -f / var / log / syslog on the gateway will not show any activity. (If I can distinguish that from my functioning node)

TD-er commented:
There is also a console log of the gateway it self on TTN.
That data window does also show any join packets received.
I don't know what you can see on the logs at Linux level on the gateway.

TD-er commented:
What is the distance between the node and the gateway?
I see I set the TX output power to `1` which equals -0.6 dBm.
I will make it also adjustable for testing.
It looks like you set it in your other node to "27" (whatever LMIC translates it to) and in your screenshot your reception RSSI was -90-ish.
So maybe if I increase the TX power you are able to connect?

TD-er commented:
Hmm, I really don't know what to believe here regarding the TX power.
The [TTN library claims](https://github.com/TheThingsNetwork/arduino-device-lib/blob/e3115ccfc1978e0c47a15410690be210bdc38936/src/TheThingsNetwork.h#L22) it should be 14 dBm (the max) and the command reference datasheet mentions -0.6 dBm in a table.

Is it possible to exchange antennas of the working node and the one you're using here on this node?
I would not exchange it the other way around, since it may destroy the transmitter of the other node if it is a defective antenna (or wrong frequency)

TD-er committed:
[LoRa TTN] Make settings configurable & add support ABP

TD-er commented:
[Test build with ABP support](https://www.dropbox.com/s/0at9fkd2422nwlc/test_core_252_ESP8266_4M_LoRa_testBuild_SupportABP.rar?dl=0)

All settings should now be configurable:
- OTAA (can leave Device EUI empty to use HW DevEUI)
- ABP
- Select Frequency plan
- Spread Factor
- SW-serial and HW-serial
- Baudrate (2400 - 115200, but for SW-serial I would stick to 9600)

I added some checks in the auto baudrate detector, so if no module is found it will not hang for too long (10 sec at most)

micropet commented:
I have my Nodes in the basement on a 10 cm long 868 MHz stub antenna in operation.

The gateway is also in the basement, but is connected to an omnidirectional antenna on the roof.

With that I can bridge a few kilometers outside.

The small power should be enough. I have to come only 2 floors to the roof.

I suspect soon that my module is broken. We tried almost everything.

I can connect an antenna to the gateway, that's the distance only 2 meters.

The syslog only provides something like:

Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: # RF packets received by concentrator: 1
Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: # CRC_OK: 100.00%, CRC_FAIL: 0.00%, NO_CRC: 0.00%
Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: # RF packets forwarded: 1 (26 bytes)
Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: # PUSH_DATA datagrams sent: 2 (554 bytes)
Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: # PUSH_DATA acknowledged: 0.00%
Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: ### [DOWNSTREAM] ###
Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: # PULL_DATA sent: 2 (0.00% acknowledged)
Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: # PULL_RESP(onse) datagrams received: 0 (0 bytes)
Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: # RF packets sent to concentrator: 0 (0 bytes)
Aug 12 21:27:18 ttn-gateway ttn-gateway[437]: # TX errors: 0

micropet commented:
The gateway antenna is now only 2 meters away from the node.

Now I get the error message
"Last Command Error: denied"

TD-er commented:
This "denied" is that in OTAA or ABP?

micropet commented:
Well, I found an RN2483 in an old node.
The node worked before.

It does not work under ESPEasy.

Still:
Last Command Error: denied

In OTAA

What could that be?

TD-er commented:
Checklist:
- Leave DevEUI empty in ESPeasy
- Make sure to topy the reported DevEUI at the bottom of the page when creating OTAA credentials for the new device.
- No spaces or other characters in the input fields.
- Disable Frame Counter Checks (on TTN console page in the settings of the device you just created)

micropet commented:
DevEUI? Where to find it?

I have only AppEUI and AppKey.
I have copied AppEUI and AppKey a few times, there are also no Spaces.

Frame Counter Check ist disabled.

TD-er commented:
> DevEUI? Where to find it?
![image](https://user-images.githubusercontent.com/3751318/62896664-ccf32c00-bd51-11e9-94b7-bbd0dd92c4b5.png)

uzi18 commented on src/ESPEasy-Globals.h:
```diff
@@ -340,17 +340,24 @@ void check_size() {
+#define CONTROLLER_LWT_CONNECT_MESSAGE      16
+#define CONTROLLER_LWT_DISCONNECT_MESSAGE   17
+#define CONTROLLER_TIMEOUT                  18
+#define CONTROLLER_ENABLED                  19  // Keep this as last, is used to loop over all parameters
```
what about enum here? should be easier to modify later

TD-er replied:
Yeah, I guess you're right, then the compiler also can warn if you forget some in a switch statement

micropet commented:
OTAA DevEUI: | 0004A30B001E7DF6 in ESPEasy is not the same as 
Device EUI        0073DB3B4914E9D6 in TTN

I do not have the Source Code

TD-er commented:
![image](https://user-images.githubusercontent.com/3751318/62896957-6de1e700-bd52-11e9-96f0-8f214eda1f1e.png)

You have to fill it in when creating (or changing) a device in TTN console

TD-er commented:
You can also use the one on the TTN page and fill it in in the ESPeasy settings.

micropet commented:
Oh crap.

I did not know anymore that you have to enter the OTAA DevEUI: at TTN.
Oh, sorry.

In the ESPeasy settings i can not make changes, because i have not the Source Code for it.

TD-er commented:
![image](https://user-images.githubusercontent.com/3751318/62897219-198b3700-bd53-11e9-9e10-9a6ef7dc7c90.png)
These have to match for OTAA.
So either you change them in TTN console, or in the settings of ESPeasy.

micropet commented:
It works now !!!.

![TTN9](https://user-images.githubusercontent.com/2838584/62897296-46d7e500-bd53-11e9-8504-fec7627ddb11.JPG)

TD-er commented:
Which module?

Oh and YAYYYY!!!!!!!

micropet commented:
The new Module, but the old will be also OK i think.

Should not the payload be displayed in legible form, as with my old nodes?

TD-er commented:
For that you need to add the decoder.js and the converter.js

But if you don't want to mess up your old config, I guess you may want to add a new "application"

micropet commented:
I have made a new application und copied the Decoder. But later i deleted the application.
I will copie die 2 Files. Just a moment....

micropet commented:
Aha, that's better


![TTN10](https://user-images.githubusercontent.com/2838584/62898002-c3b78e80-bd54-11e9-8130-2d4839149670.JPG)

micropet commented:
A BME280  and a Thermosensor MLX90614 sends data.

![TTN11](https://user-images.githubusercontent.com/2838584/62898232-4c362f00-bd55-11e9-8794-551400ec186d.JPG)

TD-er commented:
Cool :)

I'm still working on the decoder, but I guess that can be done in a later commit.
I will merge this one now, since it has a lot of changed files and I don't want to be in the way of other PRs which may cause lots of merge issues.

micropet commented:
Such a stupid mistake of mine.
Of course, I also had to do that 2 years ago with my modules.
But forget everything.

Thank you for your patience.

micropet commented:
Yes merge it !!

Good work.

I need to go to bed soon

TD-er committed:
[LoRa TTN] Add documentation stubs for C018-LoRaTTN and C017-Zabbix
